EIPS/eip-00.md

```diff
@@ -0,0 +1,789 @@
+| AA_NONCE_MANAGER      | TODO                                                                            |
+| AA_BASE_GAS_COST      | 15000                                                                           |
+| AA_ECRECOVER_COST     | 6000                                                                            |
+| VERSION               | 1                                                                               |
+| MAGIC_VALUE_SENDER    | 0xbf45c166  // keccak256("validateTransaction(uint256,bytes32,bytes)")          |
+| MAGIC_VALUE_PAYMASTER | 0xe0e6183a  // keccak256("validatePaymasterTransaction(uint256,bytes32,bytes)") |
+| MAX_CONTEXT_SIZE      | 65536                                                                           |
+
+### New Transaction Type
+
+A new [EIP-2718](./eip-2718) transaction with type AA_TX_TYPE is introduced. Transactions of this type are referred to as
+“AA transactions”. Their payload should be interpreted as:
+
+```
+
+0x04 || 0x00 || rlp([
+  chainId, sender, nonce, builderFee,
+  callData, paymasterData, deployerData,
+  maxPriorityFeePerGas, maxFeePerGas,
+  validationGasLimit, paymasterGasLimit, callGasLimit,
+  accessList, signature
+])
+
+```
+
+The base gas cost of this transaction is set to AA_BASE_GAS_COST instead of 21000 to reflect the lack of “intrinsic”
+ECDSA and signature.
+
+If `paymasterData` is specified, its first 20 bytes contain the address of a `paymaster` contract.
+
+If `deployerData` is specified, its first 20 bytes contain the address of a `deployer` contract.
+
+### Non-sequential nonce support
+
+Before EIP-9999, for accounts with associated code (smart contracts), the account nonce is only used and incremented
+when the account executes the `CREATE` (`0xf0`) opcode.
+
+However, with Smart Contract Accounts this creates a bottleneck for some use-cases.
+For example, an account that is operated by multiple participants simultaneously will require these participants
+to coordinate their transactions to avoid invalidating each other.
+
+Another example when this can also be a limitation is a case where there are separate execution flows.
+A configuration change may require multiple participants to co-sing a transaction but a regular operation does not.
+With sequential nonces, all operations will have to be halted until the configuration change is executed.
+
+To address it we propose an introduction of a separate 2-dimensional nonce used when contracts initiate a transaction.
+
+The `nonce` parameter of the transaction is to be interpreted as `uint192 key || uint64 seq` value.
+The contract account nonce is then defined as a mapping `address account => uint192 key => uint64 seq`.
+This approach guarantees unique transaction nonce and hash but removes the requirement of nonce being sequential
+numbers.
+
+This `nonce` is exposed to the EVM in a `NonceManager` pre-deployed contract located at the AA_NONCE_MANAGER address.
```

Comment:
There is a small amount of precedence with 4788 to not have "pre-deploys", (i.e. irregular state transitions). You may consider deploying the nonce manager as a normal contract and "enshrine" it's address instead.

Reply:
> There is a small amount of precedence with 4788 to not have "pre-deploys", (i.e. irregular state transitions). You may consider deploying the nonce manager as a normal contract and "enshrine" it's address instead.

We were planning to deploy it and force a certain deployment address, which is what EIP-4788 used to do with `HISTORY_STORAGE_ADDRESS`, but it was changed on Aug 24 to "A special synthetic address is generated by working backward from the desired deployment transaction".  We could do that, but there's a downside that EIP-4788 doesn't have to deal with.  This EIP is going to be used on different chains, whereas EIP-4788 only lives on L1 mainnet.  Some chains can't generate the same deployment address, such as zkSync.  It means we'll have different addresses on different chains.  

The wallet needs to read the account's current nonce for the given key from this contract, and we wouldn't want wallets to maintain a chain->NonceManager list.  We can solve it in two ways:
1. Use a fixed address as the EIP suggests, which wallets always read from.
2. RPC nodes of each network will abstract the address by adding an `eth_getNonceTxType4(account,key)` method.

Which one is better in your opinion?